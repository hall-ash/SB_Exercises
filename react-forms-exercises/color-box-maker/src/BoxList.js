import React, {useState} from 'react';
import './BoxList.css';
import NewBoxForm from './NewBoxForm';
import Box from './Box';
import { v4 as uuid } from 'uuid';

const BoxList = () => {
  const [boxes, setBoxes] = useState([]);

  const addBox = boxTraits => {
    const newBox = { ...boxTraits, id: uuid(), }
    setBoxes(boxes => [...boxes, newBox]);
  };

  const deleteBox = id => {
    const newBoxes = boxes.filter(box => box.id !== id);
    setBoxes(newBoxes);
  };

  const boxComponents = boxes.map(({bgColor, width, height, id}) => {
    return (
    <Box key={id} 
         bgColor={bgColor} 
         width={width} 
         height={height} 
         deleteBox={() => deleteBox(id)}/>);
  });

  return (
    <div>
      <h2>BoxList: </h2>
      {boxes.length ? <ul>{boxComponents}</ul> : <p>Box list is empty</p>}
      <h3>Add a new box:</h3>
      <NewBoxForm addBox={addBox}/>
    </div>
  );
};

export default BoxList;