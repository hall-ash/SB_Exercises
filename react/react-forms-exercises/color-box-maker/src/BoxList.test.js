import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import BoxList from './BoxList';

// add red box of height 300px, width 200px
const addBox = boxlist => {
  const { getByLabelText, queryByText } = boxlist;
  const bgColorInput = getByLabelText('Background color:');
  const widthInput = getByLabelText('Width:');
  const heightInput = getByLabelText('Height:');
  const addBtn = queryByText('Add');
  
  fireEvent.change(bgColorInput, { target: { value: 'red' } });
  fireEvent.change(widthInput, { target: { value: 200 } });
  fireEvent.change(heightInput, { target: { value: 300 } });

  fireEvent.click(addBtn);
};

test('renders without crashing', () => {
  render(<BoxList />);
});

test('matches snapshot', () => {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

test('contains list title', () => {
  const { getByText } = render(<BoxList />);
  const title = getByText('BoxList:');
  expect(title).toBeVisible();
});

test('contains new box form title', () => {
  const { getByText } = render(<BoxList />);
  const title = getByText('Add a new box:');
  expect(title).toBeVisible();
});

test('contains empty list message', () => {
  const { getByText } = render(<BoxList />);
  const msg = getByText('Box list is empty');
  expect(msg).toBeVisible();
});

test('add new box', () => {
  
  const boxlist = render(<BoxList />);
  const { queryByText, getAllByDisplayValue } = boxlist;
 
  // no boxes
  expect(queryByText('Box list is empty')).toBeInTheDocument();
  expect(queryByText('X')).not.toBeInTheDocument();

  // add red box of height 300px, width 200px
  addBox(boxlist)

  const deleteBtn = queryByText('X');
  expect(deleteBtn).toBeInTheDocument();
  expect(deleteBtn.previousSibling).toHaveStyle(`
    background-color: red;
    height: 300px;
    width: 200px;
  `);
  
  // expect form to be cleared (form has 3 inputs)
  expect(getAllByDisplayValue('')).toHaveLength(3);

});

test('delete box', () => {
  const boxlist = render(<BoxList />);
  const { queryByText } = boxlist;

  // add red box of height 300px, width 200px
  addBox(boxlist);

  const deleteBtn = queryByText('X');
  fireEvent.click(deleteBtn);
  expect(deleteBtn).not.toBeInTheDocument();
});