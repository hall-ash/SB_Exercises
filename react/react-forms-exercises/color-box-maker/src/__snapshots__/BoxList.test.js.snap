// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      BoxList: 
    </h2>
    <p>
      Box list is empty
    </p>
    <h3>
      Add a new box:
    </h3>
    <form>
      <label
        for="bgColor"
      >
        Background color:
      </label>
      <input
        id="bgColor"
        name="bgColor"
        type="text"
        value=""
      />
      <br />
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        type="text"
        value=""
      />
      <br />
      <label
        for="height"
      >
        Height:
      </label>
      <input
        id="height"
        name="height"
        type="text"
        value=""
      />
      <br />
      <button>
        Add
      </button>
    </form>
  </div>
</DocumentFragment>
`;
